File: CPP/torus.cpp
#include "torus.h"

// =================================================================================================

void CTorus::Create(int quality, float width, float height) {
    m_quadTexCoords = { CTexCoord(0,0), CTexCoord(1,0), CTexCoord(1,1), CTexCoord(0,1) };
    m_vertexCount = CreateVertices(quality, width, height);
    // we now have four vertex rings of m_vertexCount vertices each in the vertex buffer:
    // @ 0: lower outer vertices
    // @ vertexCount: lower inner vertices
    // @ 2 * vertexCount: upper outer vertices
    // @ 3 * vertexCount: upper inner vertices
    // now create quad indices #include "that. We will need four quad stripes:
    // outer vertical, inner vertical, lower horizontal, upper horizontal
    CreateIndex();
    CreateVAO();
}


// create four vertex rings. Resolution depends on quality ()
size_t CTorus::CreateVertices(int quality, float width, float height) {
    // compute lower outer vertex ring
    size_t vertexCount = CreateCircle(4 * size_t(pow(2, quality)), 0.5f, -height / 2.0f);
    // compute lower inner vertex ring
    float r = 1.0f - width;
    for (size_t i = 0; i < vertexCount; i++) {
        CVector& v = m_vertices[i];
        m_vertices.Append(CVector(v.X() * r, v.Y(), v.Z() * r));
    }
    // compute upper vertex rings by copying the lower vertex rings and replacing their y coordinates
    float y = height / 2.0f;
    for (size_t i = 0; i < 2 * vertexCount; i++) {
        CVector& v = m_vertices[i];
        m_vertices.Append(CVector(v.X(), y, v.Z()));
    }
    return vertexCount;
}


// create circular vertex coordinates    
size_t CTorus::CreateCircle(size_t vertexCount, float r, float y) {
    // Rad = lambda a : a / 180.0 * np.pi
    for (size_t i = 0; i < vertexCount; i++) {
        float a = float (i) / float (vertexCount) * 2.0f * float (M_PI);
        m_vertices.Append(CVector(cos(a) * r, y, sin(a) * r));
    }
    return vertexCount;
}


void CTorus::CreateIndex(void) {
    // create quads between lower and upper outer rings
    CreateQuadIndex(0, m_vertexCount * 2);
    // create quads between lower and upper inner rings
    CreateQuadIndex(m_vertexCount, m_vertexCount * 3);
    // create quads between lower outer and inner rings
    CreateQuadIndex(0, m_vertexCount);
    // create quads between upper outer and inner rings
    CreateQuadIndex(m_vertexCount * 2, m_vertexCount * 3);
}


// construct vertex indices for a quad stripe
// This fully depends on the sequence of vertices in the vertex buffer
void CTorus::CreateQuadIndex(size_t i, size_t j) {
    size_t k = 0;
    while (true) {
        size_t h = k;
        k = (h + 1) % m_vertexCount;
        CArray<size_t>* a = m_indices.m_appData.Add (-1);
        *a = { h + i, h + j, k + j, k + i };
        m_texCoords.m_appData += m_quadTexCoords;
        if (k == 0)
            break;
    }
}


// =================================================================================================

File: Python/torus.py

import globals
import numpy as np

from vector import *
from texcoord import *
from vertexdatabuffers import *
from mesh import *

# =================================================================================================

class CTorus (CMesh):
    def __init__ (self, texture = None, textureNames = None, color = CVector (1,1,1)):
        super ().__init__ (GL_QUADS, texture, textureNames, GL_TEXTURE_2D, color)
        self.vertices = CVertexBuffer ()
        self.texCoords = CTexCoordBuffer ()
        self.indices = CIndexBuffer (4)


    def Create (self, quality, width, height):
        self.quadTexCoords = [CTexCoord (0,0), CTexCoord (1,0), CTexCoord (1,1), CTexCoord (0,1)]
        self.vertexCount = self.CreateVertices (quality, width, height)
        # we now have four vertex rings of self.vertexCount vertices each in the vertex buffer:
        # @ 0: lower outer vertices
        # @ vertexCount: lower inner vertices
        # @ 2 * vertexCount: upper outer vertices
        # @ 3 * vertexCount: upper inner vertices
        # now create quad indices from that. We will need four quad stripes:
        # outer vertical, inner vertical, lower horizontal, upper horizontal
        self.CreateIndex ()
        self.CreateVAO ()


    # create four vertex rings. Resolution depends on quality ()
    def CreateVertices (self, quality, width, height):
        # compute lower outer vertex ring
        vertexCount = self.CreateCircle (4 * 2 ** quality, 0.5, -height / 2)
        # compute lower inner vertex ring
        r = 1.0 - width
        for i in range (vertexCount):
            v = self.vertices [i]
            self.vertices.Append (CVector (v.x * r, v.y, v.z * r))
        # compute upper vertex rings by copying the lower vertex rings and replacing their y coordinates
        y = height / 2
        for i in range (2 * vertexCount):
            v = self.vertices [i]
            self.vertices.Append (CVector (v.x, y, v.z))
        return vertexCount


    # create circular vertex coordinates    
    def CreateCircle (self, vertexCount, r, y):
        # Rad = lambda a : a / 180.0 * np.pi
        for i in range (0, vertexCount):
            a = i / vertexCount * 2 * np.pi
            self.vertices.Append (CVector (np.cos (a) * r, y, np.sin (a) * r))
        return vertexCount


    def CreateIndex (self):
        # create quads between lower and upper outer rings
        self.CreateQuadIndex (0, self.vertexCount * 2)
        # create quads between lower and upper inner rings
        self.CreateQuadIndex (self.vertexCount, self.vertexCount * 3)
        # create quads between lower outer and inner rings
        self.CreateQuadIndex (0, self.vertexCount)
        # create quads between upper outer and inner rings
        self.CreateQuadIndex (self.vertexCount  * 2, self.vertexCount * 3)


    # construct vertex indices for a quad stripe
    # This fully depends on the sequence of vertices in the vertex buffer
    def CreateQuadIndex (self, i, j):
        k = 0
        while True:
            h = k
            k = (h + 1) % self.vertexCount
            self.indices.Append ((h + i, h + j, k + j, k + i))
            self.texCoords.data += self.quadTexCoords
            if (k == 0):
                break                

# =================================================================================================
